libz-rs-sys/src/gz.rs

```diff
@@ -2045,6 +2045,103 @@ pub unsafe extern "C-unwind" fn gzgetc_(file: gzFile) -> c_int {
+///
+/// `gzungetc` will fail if `c` is `-1`, and may fail if a character has been pushed
+/// but not read yet. If `gzungetc` is used immediately after [`gzopen`] or [`gzdopen`],
+/// at least the output buffer size of pushed characters is allowed.  (See [`gzbuffer`].)
+///
+/// The pushed character will be discarded if the stream is repositioned with
+/// [`gzseek`] or [`gzrewind`].
+///
+/// # Returns
+///
+/// - The character pushed, on success.
+/// - `-1` on failure.
+///
+/// # Safety
+///
+/// - `file`, if non-null, must be an open file handle obtained from [`gzopen`] or [`gzdopen`].
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzungetc))]
+pub unsafe extern "C-unwind" fn gzungetc(c: c_int, file: gzFile) -> c_int {
+    let Some(state) = (unsafe { file.cast::<GzState>().as_mut() }) else {
+        return -1;
+    };
+
+    // Validate the input.
+    if c < 0 {
+        return -1;
+    }
+
+    // Check that we're reading and that there's no (serious) error.
+    if state.mode != GzMode::GZ_READ || (state.err != Z_OK && state.err != Z_BUF_ERROR) {
+        return -1;
+    }
+
+    // In case this was just opened, set up the input buffer.
+    if state.how == How::Look && state.have == 0 {
+        // We have verified that `state` is valid.
+        let _ = unsafe { gz_look(state) };
+    }
+
+    /* FIXME uncomment when seek support is implemented.
+    // Process a skip request.
+    if state.seek {
+        state.seek = false;
+        if gz_skip(state, state.skip) == -1 {
+            return -1;
+        }
+    }
+     */
+
+    // If output buffer empty, put byte at end (allows more pushing).
+    if state.have == 0 {
+        state.have = 1;
+        // Safety: because `state.have` is nonzero, the `state.output` buffer has been
+        // allocated. And because the buffer's size is `state.out_size`, a pointer to
+        // `output + out_size - 1` points within the buffer.
+        state.next = unsafe { state.output.add(state.out_size - 1) };
+        // Safety: from the addition above, `state.next` currently points within the
+        // `state.output` buffer.
+        unsafe { *(state.next as *mut u8) = c as u8 };
+        state.pos -= 1;
+        state.past = false;
+        return c;
+    }
+
+    // If no room, give up (must have already done a `gzungetc`).
+    if state.have as usize == state.out_size {
+        const MSG: &str = "out of room to push characters";
+        // Safety: We have verified that `state` is valid.
+        unsafe { gz_error(state, Some((Z_DATA_ERROR, MSG))) };
+        return -1;
+    }
+
+    // Slide output data if needed and insert byte before existing data.
+    if state.next == state.output {
+        // Safety: `state.have` is at least 1 if we got this far, so `state.output` has been
+        // allocated, and `state.next` always points within the `state.output` buffer.
+        let dst = unsafe { (state.next as *mut u8).add(1) };
```

Comment:
where does this 1 come from? I'm having some trouble mapping this back to the C original

```c
    if (state->x.next == state->out) {
        unsigned char *src = state->out + state->x.have;
        unsigned char *dest = state->out + (state->size << 1);
        while (src > state->out)
            *--dest = *--src;
        state->x.next = dest;
    }
```

Reply:
I'm still not confident that the current implementation is right, especially given that the test did not flag the problem earlier. 

This bit of logic is tricky, so while I appreciate the from-first-principles reasoning, I think it would also be useful to add why/how this maps back to the C source code, and to make sure that the tests properly exercise this logic.

I also tried running the test locally but with `use libz_sys::*;` at the top, overwriting the imports with the C version of those functions, and that fails (I commented out the line using `gzbuffer`, which `libz_sys` does not export, it then fails on  `assert_eq!(unsafe { gzungetc('-' as c_int, file) }, -1);`, which apparently returns `45` using the C version.

Reply:
I think I can address why the test didn't flag the problem in the previous version: it was a performance regression, rather than a functionality bug. zlib-ng shifts the content as far right as possible, which avoids a quadratic cost if the caller invokes gzungetc repeatedly. My original implementation, which shifted just one byte to the right, had the same functional result (the character passed to gzungetc was pushed onto the front of the buffered content) but would be very inefficient if gzungetc was called repeatedly in that state.

Reply:
`gzbuffer` plays an important role in the test case: it ensures that the buffer is small enough so the various `gzread` and `gzungetc` operations can fill and drain it. The test case sets up 16 bytes of data in the buffer before the line `assert_eq!(unsafe { gzungetc('-' as c_int, file) }, -1);` to exercise the full-buffer case. But without calling `gzbuffer`, the default buffer size in zlib-ng is something like 128KB. So the `gzungetc` call on that line will succeed, returning 45 (ascii for `-`).

I'll add some more comments to the test code to document the buffer arithmetic that the tests are using.